### PR TITLE
[WIP][Proposal]Runsy CLI framework

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
**Description of the Pull Request (PR):**

INIT commit for the singularity OCI compliant engine which will be handled on a separate CLI , temporally called RunSy. (until GA)

this PR establish the CLI framework, on which start adding the sub Cmd need for OCI functionality 

By using cobra go pkg:
 - create a `root` CMD called `runsy`
 - attach to the `root` cmd the `spec` cmd
 - attach to the `spec` cmd some basic functionality for working with OCI runtime spec config.json
 - attach to the `spec` cmd some basic functionality for working with SIF
**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

File: mlocal/frags/Makefile.stub
Comment (bauerm97) [inline mlocal/frags/Makefile.stub]: This isn't quite correct. As you can see, `singularity_OBJ` refers to the dependencies of `src/cmd/singularity/cli.go`. I think you want to introduce a variable above called `runsy_OBJ := $(shell $(SOURCEDIR)/makeit/gengodep $(SOURCEDIR)/src/cmd/runsy/cli.go)` instead.

File: mlocal/frags/Makefile.stub
Comment (bauerm97) [inline mlocal/frags/Makefile.stub]: Should remove `# set cp to install` comment. Also while you're at it, can you remove that same comment from line 73?

File: src/cmd/runsy/cli/runsy.go
Comment (bauerm97) [inline src/cmd/runsy/cli/runsy.go]: I'm not the biggest fan of introducing duplicated code between the main Singularity CLI code and the OCI code. Could we consider moving these common CLI helper functions into a new package, something like `src/pkg/util/cli`?

File: src/cmd/runsy/cli/runsy.go
Comment (bauerm97) [inline src/cmd/runsy/cli/runsy.go]: Same comment about duplicating code

File: src/docs/content-runs.go
Comment (bauerm97) [inline src/docs/content-runs.go]: I believe you meant `CLI tool`?

File: src/pkg/util/oci/helper.go
Comment (bauerm97) [inline src/pkg/util/oci/helper.go]: I *think* it's possible to just do `bn := []byte("oci-runtime-spec")`

I'm also not sure that `bn` is the best name for this, is there something slightly more descriptive that can be used?

**Review by bauerm97 — DISMISSED** [review]
It's a good start, but there are a few changes I'd like to see first off !

**Comment by ArangoGutierrez** [comment]
@bauerm97 ready for a second review round

Also will appreciate views from :
 - @GodloveD On the CLI user point of view 
 - @yhcote On the SIF functionalities

File: src/docs/content-runs.go
Comment (bauerm97) [inline src/docs/content-runs.go]: Should this file maybe be named `content-runsy.go`?

File: src/docs/makeDocs/makeDocs.go
Comment (bauerm97) [inline src/docs/makeDocs/makeDocs.go]: I would prefer if we didn't use `cli2` and instead used `runsy` or something without a number. I think it looks a little weird

**Review by bauerm97 — COMMENTED** [review]
Just two small comments. Let's wait for Yannick to get his review in before we merge. In the mean time you can keep working in a branch based off of this PR so we can still make progress!

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: use defer UnloadContainer here instead.

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: The second parameter is not the number of match or if a match was found. It is the index in the descriptor table where the match is located. GetFromDescr() as opposed to GetFromDescrAll() (coming soon) will alway only return one match or complain that the search keys are too vague to find a unique match.

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: You're looking at error here but it hasn't been set in the previous paragraph of code.

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: If "d" is != nil then a descriptor is found and this is what you should test here. Moreover why didn't you create your search descriptor (descriptor) with those to fields to match (Datatype: sif.DataGenericJSON, Name: oci.ConfigSpec). Then this test would not be necessary.

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: What's the TAB for here ?

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: This paragraph is not needed if you defer and look for errors over under the LoadContainer()

File: src/cmd/runsy/cli/spec.go
Comment (yhcote) [inline src/cmd/runsy/cli/spec.go]: I think you're better off using JoinPath here instead of and barcoded "/"

File: src/pkg/util/oci/helper.go
Comment (yhcote) [inline src/pkg/util/oci/helper.go]: You are not adding anything to the SIF file, rdonly should be true here not false (second param)

File: src/pkg/util/oci/helper.go
Comment (yhcote) [inline src/pkg/util/oci/helper.go]: See comments above about how to understand and use GetFromDescr() (what descr fields to pass in and how to interpret the return values)

File: src/pkg/util/oci/helper.go
Comment (yhcote) [inline src/pkg/util/oci/helper.go]: do defer fimg.UnloadContainer() here.

File: src/pkg/util/oci/helper.go
Comment (yhcote) [inline src/pkg/util/oci/helper.go]: This should be done with defer look at comment above.

File: src/pkg/util/oci/helper.go
Comment (yhcote) [inline src/pkg/util/oci/helper.go]: Is the TAB after column an established practice? Haven't seen it before.

**Review by yhcote — DISMISSED** [review]

**Comment by jmstover** [comment]
Re-open from `development-3.0` removal

**Pull request closed without merging**
